### PR TITLE
Implement S-LSTM function

### PR DESCRIPTION
**Pull request by unnonouno**
related to #402

### CODE REVIEW

File: chainer/functions/activation/slstm.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        return gc_prev1, gc_prev2, gx1, gx2
+
+
+def slstm(c_prev1, c_prev2, x1, x2):
+
+    """S-LSTM units as an activation function.
+
+    This function implements S-LSTM unit. It is an extention of LSTM unit
+    applied to tree structures.
+    The function is applied to binary trees. Each node has two childlren nodes.
```
Comment (beam2d) [inline chainer/functions/activation/slstm.py]: children -> child

File: chainer/functions/activation/slstm.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    It gets four arguments, previous cell states :math:`c_1` and
+    :math:`c_2`, and incoming signals :math:`x_1` and :math:`x_2`.
+
+    First both input signals :math:`x_1` and :math:`x_2` are split into
+    eight arrays :math:`a_1, i_1, f_1, o_1`, and :math:`a_2, i_2, f_2, o_2`.
+    They has the same shapes along the second axis.
```
Comment (beam2d) [inline chainer/functions/activation/slstm.py]: has -> have, shapes -> shape

File: tests/chainer_tests/functions_tests/activation_tests/test_slstm.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                           cuda.to_gpu(self.x2))
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_flat_forward_gpu(self):
+        self.flat()
+        self.test_forward_gpu()
+
+    def check_backward(self, c_prev1_data, c_prev2_data, x1_data, x2_data,
+                       c_grad, h_grad):
```
Comment (beam2d) [inline tests/chainer_tests/functions_tests/activation_tests/test_slstm.py]: Update it to use `gradient_check.check_backward` if possible.

File: chainer/functions/activation/slstm.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    It gets four arguments, previous cell states :math:`c_1` and
+    :math:`c_2`, and incoming signals :math:`x_1` and :math:`x_2`.
+
+    First both input signals :math:`x_1` and :math:`x_2` are split into
+    eight arrays :math:`a_1, i_1, f_1, o_1`, and :math:`a_2, i_2, f_2, o_2`.
+    They have the same shape along the second axis.
+    It means that :math:`x_1` and :math:`x_2` 's second axis must have 4 times
+    the length of :math:`c_{1 \\text{prev}}` and :math:`c_{2 \\text{prev}}`.
+
+    The splitted input signals are corresponding to:
```
Comment (beam2d) [inline chainer/functions/activation/slstm.py]: splitted -> split

**Comment by beam2d** [comment]
LGTM!

**Comment by kurtespinosa** [comment]
Hi, does anyone know of a tree-lstm example in chainer? I can't seem to find one. Thanks!

**Pull request merged**